launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,258 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import functools
+import inspect
+import itertools
+import warnings
+
+import launch
+from launch.launch_service import LaunchService
+import launch_testing
+import pytest
+
+
+"""
+launch_testing native pytest based implementation.
+"""
+
+
+try:
+    from _pytest.python import transfer_markers
+except ImportError:  # Pytest 4.1.0 removes the transfer_marker api (#104)
+
+    def transfer_markers(*args, **kwargs):  # noqa
+        """Noop when over pytest 4.1.0"""
+        pass
+
+
+def pytest_configure(config):
+    """Inject launch_testing marker documentation."""
+    config.addinivalue_line(
+        'markers',
+        'launch_testing: '
+        'mark the test as a launch test, it will be '
+        'run using the specified launch_pad.',
+    )
+
+# Adapted from https://github.com/pytest-dev/pytest-asyncio,
+# see their license https://github.com/pytest-dev/pytest-asyncio/blob/master/LICENSE.
+@pytest.mark.tryfirst
+def pytest_pycollect_makeitem(collector, name, obj):
+    """Collect coroutine based launch tests."""
+    if collector.funcnamefilter(name) and inspect.iscoroutinefunction(obj):
+        item = pytest.Function.from_parent(collector, name=name)
+
+        # Due to how pytest test collection works, module-level pytestmarks
+        # are applied after the collection step. Since this is the collection
+        # step, we look ourselves.
+        transfer_markers(obj, item.cls, item.module)
+        item = pytest.Function.from_parent(collector, name=name)  # To reload keywords.
+
+        if 'launch_testing' in item.keywords:
+            return list(collector._genfunctions(name, obj))
+
+
+@pytest.hookimpl(hookwrapper=True, tryfirst=True)
+def pytest_fixture_setup(fixturedef, request):
+    """Set up launch service for all launch_pytest fixtures."""
+    if getattr(fixturedef.func, '_launch_pytest_fixture', False):
+        eprefix = f"When running launch_pytest fixture '{fixturedef.func.__name__}':"
+        ls = request.getfixturevalue('launch_service')
+        event_loop = request.getfixturevalue('event_loop')
+        # get the result of the launch fixture, we take advantage of other wrappers this way
+        outcome = yield
+        ret = outcome.get_result()
+        wrong_ret_type_error = (
+            f'{eprefix} return value must be either a launch description '
+            'or a launch description, locals pair'
+        )
+        ld = ret
+        if isinstance(ret, tuple):
+            assert len(ret) == 2, wrong_ret_type_error
+            ld, _ = ret
+        assert isinstance(ld, launch.LaunchDescription), wrong_ret_type_error
+        ls.include_launch_description(ld)
+        run_async_task = event_loop.create_task(ls.run_async(
+            # TODO(ivanpauno): maybe this could be configurable (?)
+            shutdown_when_idle=True
```

Comment:
This code is only executed for `@launch_testing.pytest.fixture()` fixtures, and not for anything else.
If you want to access the launch service fixture directly and you don't have a launch_testing fixture, it won't be a problem.

Maybe I'm not understanding your comment though.
We could also have a `shutdown_when_idle` argument in `@launch_testing.pytest.fixture()`

Reply:
> This code is only executed for `@launch_testing.pytest.fixture()` fixtures, and not for anything else.

Ahh, good point.  

> If you want to access the launch service fixture directly and you don't have a launch_testing fixture, it won't be a problem.

Hmm, what's putting the launch service to run in the event loop in that case? That's a prerequisite for `wrap_coroutine` to work.

> We could also have a shutdown_when_idle argument in @launch_testing.pytest.fixture()

Yeah, but it's less obvious to me why you'd want to do that in the general case (besides adding it because it's easy).

Reply:
> Hmm, what's putting the launch service to run in the event loop in that case? That's a prerequisite for wrap_coroutine to work.

We're only using `wrap_coroutine` in tests marked with `pytest.mark.launch_testing`, which requires a fixture.
For other async tests that take a launch_service fixture directly, I think the magic that `pytest-asyncio` uses would be enough (those tests will require a `pytest.mark.asyncio` decorator).
We can certaninly test a case like that.

Reply:
> We can certaninly test a case like that.

Yeah, I'd like to see how it all plays together (fixtures with different scopes, launch tests, asyncio tests that use launch_service).

Reply:
> Yeah, but it's less obvious to me why you'd want to do that in the general case (besides adding it because it's easy).

Couldn't this replace the `KeepAliveProc()` thing?
There's really no need to launch a process to keep the launch service alive.

Reply:
How does that currently work in launch_testing by the way?

Do we stop running tests when the launch_service is idle and mark them as failed?
Should do we doing something similar here?
The code here is only ensuring that the launch service starts running before the tests and it's cleaned up before the shutdown tests, but it doesn't ensure it isn't idle before finishing running the (not shutdown) tests.

Reply:
Looking at [test_runner.py](https://github.com/ros2/launch/blob/90a245b493023d4d979d0adf8c6583083e9e33ba/launch_testing/launch_testing/test_runner.py#L163), if the launch service running task finishes before the tests, we fail. 

> Couldn't this replace the KeepAliveProc() thing?

BTW https://github.com/ros2/launch/blob/90a245b493023d4d979d0adf8c6583083e9e33ba/launch_testing/launch_testing/test_runner.py#L159 :sweat_smile:

Reply:
> Looking at test_runner.py, if the launch service running task finishes before the tests, we fail.

Okay.
I'm not sure if keeping that behavior makes sense though.

> :sweat_smile:

It's even more fun knowing that the process launched by `KeepAliveProc()` was using 100% of one cpu.

Reply:
> It's even more fun knowing that the process launched by KeepAliveProc() was using 100% of one cpu.

Oh boy :see_no_evil: